system/security/securesocket/securesocket.cpp

```diff
@@ -1142,6 +1142,161 @@ const char* strtok__(const char* s, const char* d, StringBuffer& tok)
+{
+    // this routine based on code originally from:
+    // https://stackoverflow.com/questions/3810058/read-certificate-files-from-memory-instead-of-a-file-using-openssl
+
+    // can have multiple certs in buffer
+
+    BIO *cbio = BIO_new_mem_buf((void *)certbuf, -1);
+    if (!cbio)
+        return false;
+
+    STACK_OF(X509_INFO) *infostk = PEM_X509_INFO_read_bio(cbio, NULL, NULL, NULL);
+    if (!infostk)
+    {
+        BIO_free(cbio);
+        return false;
+    }
+
+    bool first = true;
+    X509_INFO *infoval;
+    for (int i=0; i<sk_X509_INFO_num(infostk); i++)
+    {
+        infoval = sk_X509_INFO_value(infostk, i);
+        if (infoval->x509)
+        {
+            // First cert is server/main cert. Remaining, if any, are intermediate certs.
+            if (first)
+            {
+                first = false;
+
+                // Set server certificate. Note that this operation increments the
+                // reference count, which means that it is okay for cleanup to free it.
+                if (!SSL_CTX_use_certificate(ctx, infoval->x509))
+                {
+                    sk_X509_INFO_pop_free(infostk, X509_INFO_free);
+                    BIO_free(cbio);
+                    return false;
+                }
+
+                if (ERR_peek_last_error() != 0)
+                {
+                    sk_X509_INFO_pop_free(infostk, X509_INFO_free);
+                    BIO_free(cbio);
+                    return false;
+                }
+
+                // Get ready to store intermediate certs, if any.
+                SSL_CTX_clear_chain_certs(ctx);
+            }
+            else
+            {
+                // Add intermediate cert to chain.
+                if (!SSL_CTX_add0_chain_cert(ctx, infoval->x509))
+                {
+                    sk_X509_INFO_pop_free(infostk, X509_INFO_free);
+                    BIO_free(cbio);
+                    return false;
+                }
+
+                // Above function doesn't increment cert reference count. NULL the
+                // reference to it in order to prevent it from being freed during cleanup.
+                infoval->x509 = NULL;
+            }
+        }
+    }
+
+    sk_X509_INFO_pop_free(infostk, X509_INFO_free);
+    BIO_free(cbio);
+    return true;
+}
+
+static bool load_verify_locations_PEM_buffer(SSL_CTX *ctx, const char *cacertbuf)
+{
+    // this routine based on code originally from:
+    // https://stackoverflow.com/questions/5052563/c-openssl-use-root-ca-from-buffer-rather-than-file-ssl-ctx-load-verify-locat
+
+    // can have multiple certs in buffer
+
+    BIO *cbio = BIO_new_mem_buf((void *)cacertbuf, -1);
+    if (!cbio)
+        return false;
+
+    STACK_OF(X509_INFO) *infostk = PEM_X509_INFO_read_bio(cbio, NULL, NULL, NULL);
+    if (!infostk)
+    {
+        BIO_free(cbio);
+        return false;
+    }
+
+    X509_STORE *store = SSL_CTX_get_cert_store(ctx);
```

Comment:
@mckellyln what I'm wondering is, because we want only our cacerts, and don't want any default cacerts to remain...

Instead of doing:

```
 X509_STORE *store = SSL_CTX_get_cert_store(ctx);
loop{
    X509_STORE_add_cert(store, infoVal->x509);
}
```
we should do?:

```
X509_STORE *store = X509_STORE_new();
loop{
    X509_STORE_add_cert(store, infoVal->x509);
}
SSL_CTX_set_cert_store(ctx, store);  //this will clear (and free) old and add new

```